### PR TITLE
[MRG] Fix several SVG closing issues

### PR DESCRIPTION
**Pull request by QuLogic**
Fix a leak where the wrong file handle was closed (maybe related to warnings in #6870? @Kojoley). Fix premature closing of underlying files, and some pickling issues.

### CODE REVIEW

File: lib/matplotlib/backends/backend_svg.py [hunk 1]
```diff
@@ -1230,8 +1234,10 @@ def _print_svg(self, filename, svgwriter, fh_to_close=None, **kwargs):
             self.figure.draw(renderer)
             renderer.finalize()
         finally:
+            if detach:
+                svgwriter.detach()
```
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py]: Hope this works on Python 2...

Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py]: Guess I'll be needing a Python 2 version.

File: lib/matplotlib/backends/backend_svg.py [hunk 2]
```diff
@@ -1235,7 +1235,11 @@ def _print_svg(self, filename, svgwriter, fh_to_close=None, detach=False,
             renderer.finalize()
         finally:
             if detach:
-                svgwriter.detach()
+                if six.PY3:
+                    svgwriter.detach()
+                else:
+                    svgwriter.reset()
+                    svgwriter.stream = io.BytesIO()
```
Comment (Kojoley) [inline lib/matplotlib/backends/backend_svg.py]: I think you are doing it not in a right way. Just `RendererSVG.writer` should be deleted in `RendererSVG.finalize`.

Closes #6870

Comment (tacaswell) [inline lib/matplotlib/backends/backend_svg.py]: The call to finalize should go it in the finally block then.

**Comment by tacaswell** [comment]
I do not fully understand the difference between `detach` and `fh_to_close`.

There may need to be a bigger change in that renderers are allowed to report that they can not be cached (as I suspect that calling `draw_artist` with the svg renderer would not work).

File: lib/matplotlib/backends/backend_svg.py (hunk 2 above)
Comment (QuLogic) [inline lib/matplotlib/backends/backend_svg.py]: This is not intended to fix #6870. It is the Python 2 equivalent of 946367f, which fixes #6926.

#6870 is fixed by 7fe62ac, because the entire cached renderer (not just this part) is not supposed to be pickled.

**Comment by Kojoley** [comment]
I previously refactored `print_svg` and `print_svgz` with context manager, but have not opened a PR with the changes. Would you like me to open a PR?

**Comment by QuLogic** [comment]
Sure, open directly to this branch and we'll include it in this PR.

**Comment by Kojoley** [comment]
It does not clearly cherry-picks here due to added detach magic.

**Comment by QuLogic** [comment]
Post a branch to your fork and I'll see; maybe I can drop the detach stuff...

**Comment by Kojoley** [comment]
https://github.com/Kojoley/matplotlib/commit/5ce372f07075830c92c971731f11c05ae0210cf1

**Comment by QuLogic** [comment]
Thanks; I've incorporated it into this PR and we can drop the confusing `fh_to_close` stuff. However, the detach stuff is still necessary to fix #6926.

**Comment by QuLogic** [comment]
Restarted the failed build, so it's good from my side.

**Comment by dopplershift** [comment]
So I'm a little rusty--do I need to backport this to master?

@QuLogic will you make sure all the issues fixed by this get closed?

**Comment by tacaswell** [comment]
@dopplershift no, we every-so-often merge v2.x into master.  

I think @QuLogic and I are currently playing chicken about who will do it next :wink: .

**Comment by QuLogic** [comment]
The issues will close automatically when the commits get into `master`.
